### PR TITLE
bpo-35105: Document that CPython accepts "invalid" identifiers

### PR DESCRIPTION
**Pull request by Windsooon**
https://bugs.python.org/issue35105

### CODE REVIEW

File: Doc/library/functions.rst [hunk 1]
```diff
@@ -1388,6 +1388,11 @@ are always available.  They are listed here in alphabetical order.
    object allows it.  For example, ``setattr(x, 'foobar', 123)`` is equivalent to
    ``x.foobar = 123``.
 
+   .. impl-detail::
+
+   The attribute name does not have to be a valid identifier, For example,
+   ``setattr(x, "1", 123)``, this is not guaranteed behaviour and may change
+   in the future.
```
Comment (terryjreedy) [inline Doc/library/functions.rst]: The second 'sentence' is not grammatically correct English.  (And no 'u' in the American spelling of 'behavior'.)  The first sentence could be seen appears to say that the behavior is more or less guaranteed.  An alternative:
"An implementation may accept attribute names that are not valid identifiers, as in ``setattr(x, "1", 123)``.  But this is not guaranteed."

**Review by terryjreedy — CHANGES_REQUESTED** [review]
That you for proposing something specific to chew on.  I think spelling, grammar, and content all need tweaking.

File: Doc/library/functions.rst (hunk 1 above)
Comment (terryjreedy) [reply Doc/library/functions.rst]: Correct spelling and grammar now, but I recommend waiting a few days for other opinions before rushing to accept my content suggestion.

**Comment by zhangyangyu** [comment]
Do we really need to mention this?

**Comment by terryjreedy** [comment]
That discussion is on the issue.

File: Doc/library/functions.rst (hunk 1 above)
Comment (Windsooon) [reply Doc/library/functions.rst]: Thank you @terryjreedy, you are a better writer than me :D. I found the document use [behavior](https://docs.python.org/3/tutorial/introduction.html#numbers) as well as [behaviour](https://docs.python.org/3/tutorial/inputoutput.html#methods-of-file-objects), I kinda confuse which one we should use.

**Review by rhettinger — CHANGES_REQUESTED** [review]
I don't think we can mark this as an implementation detail for *setattr()*.  The details are downstream and determined by the target object, not by setattr itself.

Suggested wording:

    Note, setattr() attempts to update the object with the given attr/value pair.
    Whether this succeeds and what its affect is is determined by the target object.
    If an object's class defines `__slots__`, the attribute may not be writeable.
    If an object's class defines property with a setter method, the *setattr()*
    will trigger the setter method which may or may not actually write the attribute.
    For objects that have a regular dictionary (which is the typical case), the
    *setattr()* call can make any string keyed update allowed by the dictionary
    including keys that aren't valid identifiers -- for example setattr(a, '1', 'one')
    will be the equivalent of vars()['1'] ='one'.

This issue has little to do with *setattr()* and is more related to the fact that instance dictionaries can hold any valid key.  In a way, it is no different than a user writing `a.__dict__['1'] = 'one'`.  That has always been allowed and the `__dict__` attribute is documented as writeable, so a user is also allowed to write `a.__dict__ = {'1': 'one'}.

In short, we can talk about this in the setattr() docs but it isn't really a setattr() issue.  Also, the behavior is effectively guaranteed by the other things users are allowed to do, so there is no merit in marking this as an implementation detail.  Non-identifier keys can make it into an instance dictionary via multiple paths that are guaranteed to work.

**Comment by terryjreedy** [comment]
I agree with Raymond's revision.  Note that it would replace the entire proposed addition, including the impl-detail directive.  By adding more 'meat', it makes an addition more worthwhile.

**Review by csabella — CHANGES_REQUESTED** [review]
The whole note needs to be indented for proper alignment.

File: Doc/library/functions.rst [hunk 2]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
    object allows it.  For example, ``setattr(x, 'foobar', 123)`` is equivalent to
    ``x.foobar = 123``.
 
+.. note::
+
+   setattr() attempts to update the object with the given attr/value pair.
+   Whether this succeeds and what its affect is is determined by the target object.
```
Comment (csabella) [inline Doc/library/functions.rst]: I would suggest changing `what its affect is` to `the affect it has` so that `is is` doesn't trigger future (incorrect) doc fixes.

File: Doc/library/functions.rst [hunk 3]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
...
+   If an object's class defines `__slots__`, the attribute may not be writeable.
+   If an object's class defines property with a setter method, the *setattr()*
+   will trigger the setter method which may or may not actually write the attribute.
+   For objects that have a regular dictionary (which is the typical case), the
+   *setattr()* call can make any string keyed update allowed by the dictionary
+   including keys that aren't valid identifiers -- for example setattr(a, '1', 'one')
```
Comment (csabella) [inline Doc/library/functions.rst]:
```suggestion
   including keys that aren't valid identifiers -- for example ``setattr(a, '1', 'one')``
```

File: Doc/library/functions.rst [hunk 4]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
...
+   will be the equivalent of vars()['1'] ='one'.
```
Comment (csabella) [inline Doc/library/functions.rst]:
```suggestion
   will be the equivalent of ``vars()['1'] ='one'``.
```

File: Doc/library/functions.rst [hunk 5]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
    object allows it.  For example, ``setattr(x, 'foobar', 123)`` is equivalent to
    ``x.foobar = 123``.
 
+.. note::
+
+   setattr() attempts to update the object with the given attr/value pair.
+   Whether this succeeds and what its affect is is determined by the target object.
+   If an object's class defines `__slots__`, the attribute may not be writeable.
+   If an object's class defines property with a setter method, the *setattr()*
```
Comment (csabella) [inline Doc/library/functions.rst]:
```suggestion
   If an object's class defines :class:`property` with a setter method, the *setattr()*
```

File: Doc/library/functions.rst (hunk 2 above)
Comment (csabella) [reply Doc/library/functions.rst]: Actually, I think `affect` should be `effect`.

**Comment by JulienPalard** [comment]
Hi @Windsooon thanks for this PR! Would you have some time to integrate Cheryl comments?

**Comment by Windsooon** [comment]
I just updated it. Thank you @csabella

File: Doc/library/functions.rst [hunk 6]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
    object allows it.  For example, ``setattr(x, 'foobar', 123)`` is equivalent to
    ``x.foobar = 123``.
 
+.. note::
+
+   setattr() attempts to update the object with the given attr/value pair.
```
Comment (JulienPalard) [inline Doc/library/functions.rst:1393]:
```suggestion
   ``setattr()`` attempts to update the object with the given attr/value pair.
```

File: Doc/library/functions.rst [hunk 7]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
...
+   Whether this succeeds and the effect it has determined by the target object.
+   If an object's class defines `__slots__`, the attribute may not be writeable.
```
Comment (JulienPalard) [inline Doc/library/functions.rst:1395]:
```suggestion
   If an object's class defines ``__slots__``, the attribute may not be writeable.
```

File: Doc/library/functions.rst [hunk 8]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
...
+   If an object's class defines :class:`property` with a setter method, the *setattr()*
+   will trigger the setter method which may or may not actually write the attribute.
+   For objects that have a regular dictionary (which is the typical case), the
+   *setattr()* call can make any string keyed update allowed by the dictionary
+   including keys that aren't valid identifiers -- for example ``setattr(a, '1', 'one')``
+   will be the equivalent of ``vars()['1'] ='one'``.
```
Comment (JulienPalard) [inline Doc/library/functions.rst:1401]:
```suggestion
   will be the equivalent of ``vars()['1'] = 'one'``.
```

**Review by JulienPalard — CHANGES_REQUESTED** [review]
LGTM with nits (all can be accepted right from github).

**Comment by csabella** [comment]
@Windsooon, please address the review comments from Julien.  Thanks!

**Comment by AlexWaygood** [comment]
Changes were requested on this PR by a core dev over two years ago, but have not been applied. There is also now a merge conflict. I am therefore closing this PR.

@Windsooon, if you're still interested in working on this, feel free to open a new PR. Alternatively, ping me, and I'll happily reopen the PR. Thanks! 🙂

**Comment by Windsooon** [comment]
@AlexWaygood I'm so sorry for missing the notification from CPython, I will catch up in a few weeks.

**Comment by AlexWaygood** [comment]
@Windsooon no worries!

File: Doc/library/functions.rst (hunk 6 above)
Comment (slateny) [inline Doc/library/functions.rst:1393]: (in addition to the above)

```suggestion
   setattr() attempts to update the object with the given attribute/value pair.
```

or

```suggestion
   setattr() attempts to update the object with the given attribute-value pair.
```

File: Doc/library/functions.rst [hunk 9]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
    object allows it.  For example, ``setattr(x, 'foobar', 123)`` is equivalent to
    ``x.foobar = 123``.
 
+.. note::
+
+   setattr() attempts to update the object with the given attr/value pair.
+   Whether this succeeds and the effect it has determined by the target object.
```
Comment (slateny) [inline Doc/library/functions.rst:1394]:
```suggestion
   Whether this succeeds and the effect it has is determined by the target object.
```

File: Doc/library/functions.rst [hunk 10]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
...
+   If an object's class defines `__slots__`, the attribute may not be writeable.
+   If an object's class defines :class:`property` with a setter method, the *setattr()*
```
Comment (slateny) [inline Doc/library/functions.rst:1396]: `*setattr()*` -> `` `setattr()` `` here and below (I think * is commonly used for parameters)

File: Doc/library/functions.rst [hunk 11]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
...
+   will trigger the setter method which may or may not actually write the attribute.
+   For objects that have a regular dictionary (which is the typical case), the
+   *setattr()* call can make any string keyed update allowed by the dictionary
+   including keys that aren't valid identifiers -- for example ``setattr(a, '1', 'one')``
```
Comment (slateny) [inline Doc/library/functions.rst:1400]: Would you happen to have an example of this behavior? I'm not quite following what it's doing (how `a` turns into `vars()`)

File: Doc/library/functions.rst [hunk 12]
```diff
@@ -1388,6 +1388,17 @@ are always available.  They are listed here in alphabetical order.
...
 
+.. note::
+
+   setattr() attempts to update the object with the given attr/value pair.
+   Whether this succeeds and the effect it has determined by the target object.
+   If an object's class defines `__slots__`, the attribute may not be writeable.
+   If an object's class defines :class:`property` with a setter method, the *setattr()*
+   will trigger the setter method which may or may not actually write the attribute.
+   For objects that have a regular dictionary (which is the typical case), the
+   *setattr()* call can make any string keyed update allowed by the dictionary
```
Comment (slateny) [inline Doc/library/functions.rst:1399]:
```suggestion
   *setattr()* call can make any string-keyed update allowed by the dictionary
```

**Comment by AlexWaygood** [comment]
I think this PR has now been superseded by #96454
